source/common/http/conn_manager_impl.cc

```diff
@@ -598,6 +599,17 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
   // called with end_stream=true.
   maybeEndDecode(end_stream);
 
+  if (connection_manager_.config_.validateHeaders()) {
```

Comment:
I do just want to call out this solves your problem because your client is using HTTP/1.  If they use HTTP/2 with invalid characters, nghttp2 would reject the request before we got to the HCM decodeHeaders call.
I still think it's a good change, I just want to make sure the limitations are on your radar :-)